### ISSUE TITLE
GL-INSTANCE-ID not defined

### ISSUE THREAD

**Issue by kiran-kp** [cbaggers/varjo#124]
Apologies if I'm missing something obvious, but I get this error when trying to build the triangle instancing example. I barely understand common lisp so my sleuthing didn't help much. I'm using the quicklisp release of varjo.

```
Varjo: Symbol 'MY-PROJ::GL-INSTANCE-ID' is unidentified.
   [Condition of type VARJO-CONDITIONS:SYMBOL-UNIDENTIFIED]
```

**Comment by cbaggers** [comment]
Hi there, sorry to hear about the issue. Did you follow the steps [here](https://github.com/cbaggers/cepl.examples/blob/master/examples/how-to-run-these-examples.md)?

**Comment by kiran-kp** [comment]
Not quite. I have the incorrect version of assimp installed (v4.0) so I can't load the examples project directly. I copied the instancing example specifically to another empty cepl project that I created.

I've verified that the setup is correct by running simpler code like just rendering a triangle and messing around with the triangle's params. Any use of 'gl-instance-id' though results in the compilation error I posted above.

**Comment by cbaggers** [comment]
odd, which packages are you using in that project?

**Comment by kiran-kp** [comment]
This is what my asd file contains:
```
:depends-on (#:cepl
             #:cepl.sdl2
             #:swank
             #:livesupport
             #:cepl.skitter.sdl2
             #:dirt)
```

Poking around in slime autocompletion I do see vari.glsl:gl-instance-id.

The code I'm trying to compile is a straight copy of [this](https://github.com/cbaggers/cepl.examples/blob/master/examples/instance-array-triangles.lisp)

This is the full error and backtrace I see for the compilation failure
```
Varjo: Symbol 'MY-PROJ::GL-INSTANCE-ID' is unidentified.
   [Condition of type VARJO-CONDITIONS:SYMBOL-UNIDENTIFIED]
```
```
  0: (VARJO.INTERNALS::RAISE-TEST-TRANSLATE-ERROR (#<VARJO-CONDITIONS:SYMBOL-UNIDENTIFIED {1005CEDCA3}> #<VARJO-CONDITIONS:SYMBOL-UNIDENTIFIED {1005ECD463}> #<VARJO-CONDITIONS:SYMBOL-UNIDENTIFIED {1005FD3D..
  1: (VARJO.INTERNALS::%TEST-TRANSLATE-RAISING #<CLOSURE (LABELS VARJO.INTERNALS::GET-STAGE :IN VARJO.INTERNALS:TEST-TRANSLATE-FUNCTION-SPLIT-DETAILS) {10073ADB8B}> (:VERTEX :TESSELLATION-CONTROL :TESSELLA..
  2: (CEPL.PIPELINES::%TEST-&-UPDATE-SPEC #<CEPL.PIPELINES::GPU-FUNC-SPEC {10073AC503}>)
  3: ((SB-C::TOP-LEVEL-FORM (CEPL.PIPELINES::%TEST-&-UPDATE-SPEC CEPL.PIPELINES::GPU-FUNC-SPEC))) [toplevel]
  4: (SB-FASL::LOAD-FASL-GROUP #S(SB-FASL::FASL-INPUT :STREAM #<SB-SYS:FD-STREAM for "file /home/kiran/.cache/common-lisp/sbcl-1.4.0-linux-x64/home/kiran/sandbox/cepl/my-proj/my-proj.fasl" {1006FD6A73}> :T..
  5: (SB-FASL::LOAD-AS-FASL #<SB-SYS:FD-STREAM for "file /home/kiran/.cache/common-lisp/sbcl-1.4.0-linux-x64/home/kiran/sandbox/cepl/my-proj/my-proj.fasl" {1006FD6A73}> NIL NIL)
  6: ((FLET SB-FASL::THUNK :IN LOAD))
  7: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<CLOSURE (FLET SB-FASL::THUNK :IN LOAD) {7FFFF4964BFB}> #<SB-SYS:FD-STREAM for "file /home/kiran/.cache/common-lisp/sbcl-1.4.0-linux-x64/home/kiran/sandbox/cepl/my-p..
  8: ((FLET SB-FASL::LOAD-STREAM :IN LOAD) #<SB-SYS:FD-STREAM for "file /home/kiran/.cache/common-lisp/sbcl-1.4.0-linux-x64/home/kiran/sandbox/cepl/my-proj/my-proj.fasl" {1006FD6A73}> T)
  9: (LOAD #P"/home/kiran/.cache/common-lisp/sbcl-1.4.0-linux-x64/home/kiran/sandbox/cepl/my-proj/my-proj.fasl" :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST T :EXTERNAL-FORMAT :DEFAULT)
 10: (UIOP/UTILITY:CALL-WITH-MUFFLED-CONDITIONS #<CLOSURE (LAMBDA NIL :IN UIOP/LISP-BUILD:LOAD*) {1006FD572B}> ("Overwriting already existing readtable ~S." #(#:FINALIZERS-OFF-WARNING :ASDF-FINALIZERS)))
 11: ((SB-PCL::EMF ASDF/ACTION:PERFORM) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/LISP-ACTION:CL-SOURCE-FILE "my-proj" "my-proj">)
 12: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS (ASDF/LISP-ACTION:LOAD-OP ASDF/LISP-ACTION:CL-SOURCE-FILE)) #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/LISP-ACTION:CL-SOURCE-FILE "my-proj" "my-proj">)..
 13: ((:METHOD ASDF/ACTION:PERFORM-WITH-RESTARTS :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/LISP-ACTION:CL-SOURCE-FILE "my-proj" "my-proj">) [fast-method]
 14: ((:METHOD ASDF/PLAN:PERFORM-PLAN (LIST)) ((#<ASDF/LISP-ACTION:PREPARE-OP > . #1=#<ASDF/SYSTEM:SYSTEM "uiop">) (#2=#<ASDF/LISP-ACTION:COMPILE-OP > . #1#) (#<ASDF/LISP-ACTION:LOAD-OP > . #1#) (#2# . #<A..
 15: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
 16: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) ((#<ASDF/LISP-ACTION:PREPARE-OP > . #1=#<ASDF/SYSTEM:SYSTEM "uiop">) (#2=#<ASDF/LISP-ACTION:COMPILE-OP > . #1#) (#<ASDF/LISP-ACTION:LOAD-OP > . #1#) (#2# ..
 17: ((FLET SB-C::WITH-IT :IN SB-C::%WITH-COMPILATION-UNIT))
 18: ((:METHOD ASDF/PLAN:PERFORM-PLAN :AROUND (T)) #<ASDF/PLAN:SEQUENTIAL-PLAN {1005BE9CD3}> :VERBOSE NIL) [fast-method]
 19: ((:METHOD ASDF/OPERATE:OPERATE (ASDF/OPERATION:OPERATION ASDF/COMPONENT:COMPONENT)) #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/SYSTEM:SYSTEM "my-proj"> :VERBOSE NIL) [fast-method]
 20: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/SYSTEM:SYSTEM "my-proj"> :VERBOSE NIL)
 21: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
 22: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) #<ASDF/LISP-ACTION:LOAD-OP :VERBOSE NIL> #<ASDF/SYSTEM:SYSTEM "my-proj"> :VERBOSE NIL) [fast-method]
 23: ((SB-PCL::EMF ASDF/OPERATE:OPERATE) #<unused argument> #<unused argument> ASDF/LISP-ACTION:LOAD-OP "my-proj" :VERBOSE NIL)
 24: ((LAMBDA NIL :IN ASDF/OPERATE:OPERATE))
 25: (ASDF/CACHE:CALL-WITH-ASDF-CACHE #<CLOSURE (LAMBDA NIL :IN ASDF/OPERATE:OPERATE) {1005BE0E9B}> :OVERRIDE NIL :KEY NIL)
 26: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP "my-proj" :VERBOSE NIL) [fast-method]
 27: ((:METHOD ASDF/OPERATE:OPERATE :AROUND (T T)) ASDF/LISP-ACTION:LOAD-OP "my-proj" :VERBOSE NIL) [fast-method]
 28: (ASDF/OPERATE:LOAD-SYSTEM "my-proj" :VERBOSE NIL)
 29: (QUICKLISP-CLIENT::CALL-WITH-MACROEXPAND-PROGRESS #<CLOSURE (LAMBDA NIL :IN QUICKLISP-CLIENT::APPLY-LOAD-STRATEGY) {1005A761DB}>)
 30: (QUICKLISP-CLIENT::AUTOLOAD-SYSTEM-AND-DEPENDENCIES "my-proj" :PROMPT NIL)
 31: ((:METHOD QL-IMPL-UTIL::%CALL-WITH-QUIET-COMPILATION (T T)) #<unused argument> #<CLOSURE (FLET QUICKLISP-CLIENT::QL :IN QUICKLISP-CLIENT:QUICKLOAD) {1005A68E8B}>) [fast-method]
 32: ((:METHOD QL-IMPL-UTIL::%CALL-WITH-QUIET-COMPILATION :AROUND (QL-IMPL:SBCL T)) #<QL-IMPL:SBCL {1000F0BF23}> #<CLOSURE (FLET QUICKLISP-CLIENT::QL :IN QUICKLISP-CLIENT:QUICKLOAD) {1005A68E8B}>) [fast-me..
 33: ((:METHOD QUICKLISP-CLIENT:QUICKLOAD (T)) ("my-proj") :PROMPT NIL :SILENT NIL :VERBOSE NIL) [fast-method]
 34: (QL-DIST::CALL-WITH-CONSISTENT-DISTS #<CLOSURE (LAMBDA NIL :IN QUICKLISP-CLIENT:QUICKLOAD) {1005A60C9B}>)
 35: (SB-INT:SIMPLE-EVAL-IN-LEXENV (QUICKLISP-CLIENT:QUICKLOAD "my-proj") #<NULL-LEXENV>)
 36: (EVAL (QUICKLISP-CLIENT:QUICKLOAD "my-proj"))
 37: (SWANK::EVAL-REGION "(ql:quickload \"my-proj\") ..)
 38: ((LAMBDA NIL :IN SWANK-REPL::REPL-EVAL))
 39: (SWANK-REPL::TRACK-PACKAGE #<CLOSURE (LAMBDA NIL :IN SWANK-REPL::REPL-EVAL) {1005A607BB}>)
 40: (SWANK::CALL-WITH-RETRY-RESTART "Retry SLIME REPL evaluation request." #<CLOSURE (LAMBDA NIL :IN SWANK-REPL::REPL-EVAL) {1005A6075B}>)
 41: (SWANK::CALL-WITH-BUFFER-SYNTAX NIL #<CLOSURE (LAMBDA NIL :IN SWANK-REPL::REPL-EVAL) {1005A6073B}>)
 42: (SWANK-REPL::REPL-EVAL "(ql:quickload \"my-proj\") ..)
 43: (SB-INT:SIMPLE-EVAL-IN-LEXENV (SWANK-REPL:LISTENER-EVAL "(ql:quickload \"my-proj\") ..)
 44: (EVAL (SWANK-REPL:LISTENER-EVAL "(ql:quickload \"my-proj\") ..)
 45: (SWANK:EVAL-FOR-EMACS (SWANK-REPL:LISTENER-EVAL "(ql:quickload \"my-proj\") ..)
 46: (SWANK::PROCESS-REQUESTS NIL)
 47: ((LAMBDA NIL :IN SWANK::HANDLE-REQUESTS))
 48: ((LAMBDA NIL :IN SWANK::HANDLE-REQUESTS))
 49: (SWANK/SBCL::CALL-WITH-BREAK-HOOK #<FUNCTION SWANK:SWANK-DEBUGGER-HOOK {21DC09EB}> #<CLOSURE (LAMBDA NIL :IN SWANK::HANDLE-REQUESTS) {1002F3802B}>)
 50: ((FLET SWANK/BACKEND:CALL-WITH-DEBUGGER-HOOK :IN "/home/kiran/.emacs.d/elpa/slime-20171106.1331/swank/sbcl.lisp") #<FUNCTION SWANK:SWANK-DEBUGGER-HOOK {21DC09EB}> #<CLOSURE (LAMBDA NIL :IN SWANK::HAND..
 51: (SWANK::CALL-WITH-BINDINGS ((*STANDARD-INPUT* . #1=#<SWANK/GRAY::SLIME-INPUT-STREAM {1002E47B23}>) (*STANDARD-OUTPUT* . #2=#<SWANK/GRAY::SLIME-OUTPUT-STREAM {1002F17973}>) (*TRACE-OUTPUT* . #2#) (*ERR..
 52: (SWANK::HANDLE-REQUESTS #<SWANK::MULTITHREADED-CONNECTION {1002559763}> NIL)
 53: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
 54: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
 55: ((FLET "WITHOUT-INTERRUPTS-BODY-1" :IN SB-THREAD::CALL-WITH-MUTEX))
 56: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {7FFFF4966D1B}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THR..
 57: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "repl-thread" RUNNING {1002F2FFA3}> NIL #<CLOSURE (LAMBDA NIL :IN SWANK-REPL::SPAWN-REPL-THREAD) {1002F2FF4B}> (#<SB-THREAD:THREAD "re..
 58: ("foreign function: call_into_lisp")
 59: ("foreign function: new_thread_trampoline")
```

**Comment by kiran-kp** [comment]
I'm still discovering parts of slime/swank so I just discovered I can inspect locals in the backtrace.

It looks like gl-instance-id isn't being resolved correctly for some reason. I qualified it as `vari.glsl:gl-instance-id` and I was able build and run the example. I guess the problem was less of a varjo issue and something wrong with the way I set up the project in that's it's not resolving symbols correctly. I'll figure out common lisp eventually :smiley: 

Thanks for the quick responses. CEPL is a very exciting new toy and some awesome tech!
